### ISSUE TITLE
Pull through caching 

### ISSUE THREAD

**Issue by jsheetzati** [Azure/acr#599]
**What is the problem you're trying to solve**
With Docker licensing and rate limiting changes, we are interested in using our ACR instance to mirror our base images. The current solution seems to be manually importing base images into ACR and creating an ACR task to pull in base image updates. 

**Describe the solution you'd like**
Ability to use ACR as a pull through cache similar to what was added to AWS. For us, this would help with the Docker changes and save us work from manually importing and updating the base images. 

https://aws.amazon.com/blogs/aws/announcing-pull-through-cache-repositories-for-amazon-elastic-container-registry/

**Comment by logamanig** [comment]
It would be much better to support multiple public registries, including major ones ghcr.io, quay.io, docker.io and gcr.io as Images from these major registries needs to be imported to private registry for image scanning and firewall restriction reasons

**Comment by tssgery** [comment]
Is there any update on this, it'd sure come in handy and solve some very large problems for me.

**Comment by squillace** [comment]
Yes, @tssgery apologies on the slow delivery here. We had to work through some authentication issues as well as other challenges, but we are getting ready for preview by the end of January. We regret the delay, of course. 

Do you have a scenario you can describe here? I'd love to ensure it'll be picked up by what we're doing here.

**Comment by mmuffins** [comment]
Something that we want(ed) to do is a repository for container images that we can then pass on to other environments.

The idea was to create a single central acr used as repository for internal golden images, as well as cache for external container registries like dockerhub or mcr.
In addition to the central registry, each project/dev environment/etc. has an additional local registry that's configured to pull everything from the central one.

That way dev teams can upload their in-development images to their own isolated container registry while still having access to approved company images and images from external sources.

In addition, that makes it easier to control what external images are being used, e.g. by configuring the central repository to only allow pulling from verified uploaders.

Nothing that couldn't technically be done with separate registries, but having just a single source of truth to moderate makes things much easier.

**Comment by squillace** [comment]
This is, essentially, precisely one thing you'll get in late January, @mmuffins. You'll get other nice features as well. Keep an eye out for the announcement.

**Comment by jkroepke** [comment]
@squillace But you do not talk about connected registries, where all images are sync into the own ACR?

**Comment by marrotte** [comment]
Hi.  Any updates on this one?

**Comment by aaronkorver** [comment]
Adding on that I'm extremely interested in this feature as well for my company.  Looking forward to the announcement!  Should we watch an additional place for the release notes?

**Comment by squillace** [comment]
You can watch here, but there will be a big blog post and a customer email announcement as well.

**Comment by jkroepke** [comment]
There already private beta for opt-in?

**Comment by squillace** [comment]
Hi all, look for our new feature announcement tomorrow. I will update this thread with links at that time. I'd really like to hear from everyone who tries it and collect as much feedback as possible, as we'll be iterating this feature rapidly to ensure we're hitting the high notes.

**Comment by hho** [comment]
As yesterday came and went without any announcement, I guess the new documentation is https://learn.microsoft.com/en-us/azure/container-registry/tutorial-registry-cache … which unfortunately doesn't help me, because it currently excludes private registries. :disappointed:

**Comment by squillace** [comment]
https://techcommunity.microsoft.com/t5/apps-on-azure-blog/announcing-public-preview-of-caching-for-acr/ba-p/3744655

**Comment by squillace** [comment]
@hho, we just had to touch one little thing, which pushed it to this morning PST.

**Comment by squillace** [comment]
@hho what do you mean, https://github.com/Azure/acr/issues/599#issuecomment-1433495512 excludes private networks? It explicitly supports private networks.

**Comment by squillace** [comment]
But "support" in this case means it's _transparent_. You do not need to enable anything directly; just create a private network ACR, set up your cache rules, and inside that private network from the ACR instance cached repo. It should work fine. 

@@@@@NOTE: You must pull once to "hydrate" the image. It doesn't arrive automatically. You are in control of pulling the images into the cache. But once there, your next pull comes directly from that ACR cached image.@@@@@@@

**Comment by hho** [comment]
@squillace Thanks for the update!

I wasn't talking about networks at all, I'm only referring to the [preview's limitation of not supporting our own registry](https://learn.microsoft.com/en-us/azure/container-registry/tutorial-registry-cache#:~:text=self%2Dhosted%20registries%20are%20on%20the%20roadmap%20but%20aren%27t%20included%20in%20this%20release).

**Comment by squillace** [comment]
Ah, @hho, I'd call them "custom" or "self-hosted registries" as there are a LOT of possibilities there. We have customers who are waiting on this support, which I expect to come rapidly now that we have the platform. Do you have a specific one in mind we should test against?

**Comment by hho** [comment]
@squillace That's great news! I'd be most interested in [JFrog Artifactory](https://jfrog.com/pricing/#devops-cloud) (currently on-prem, but they also have a cloud offering you could test against).

**Comment by squillace** [comment]
JFrog is already one of the leading candidates suggested by other customers, so we'll definitely have that in the first list of support! I'll update this when I have clearer ideas of the next wave.

**Comment by carlossg** [comment]
From the docs I guess you cannot do caching of another ACR registry (ie. in another region)?

**Comment by squillace** [comment]
not yet; ACR support will be in the next iteration. We have people who would like to chain cache instances of ACR for various departments and so on.

**Comment by jkroepke** [comment]
Happy to see quay.io and ghcr.io Support, since a lot of upstream images are not longer using Docker Hub anymore, e.g Prometheus Community.

**Comment by squillace** [comment]
> 

quay and ghcr are in line, but not yet supported. We'll get there soon!

**Comment by tssgery** [comment]
Just to throw in my opinion as well... Privately hosted JFrog Artifactory instances are our most important need now that docker.io support is in place.

**Comment by diegopasten** [comment]
Really nice to see this feature!

I'm also really happy seeing that you mention support for ghcr and quay, but is there a slightly more specific ETA for other registries? I would love to see quay and ghcr before May. Is that realistic?

**Comment by leinad87** [comment]
Nice to see this feature. Right now, you need to configure each image you want to cache with a rule (up to 50) them the system do [this](https://learn.microsoft.com/en-us/azure/container-registry/buffer-gate-public-content?tabs=azure-cli#import-images-to-an-azure-container-registry) automatically for you.

Could be possible to configure as a proxy-caché system? Trying to download the image from CR if found, otherwise search the image on external repositories and cache it in the CR for future downloads. This is available on other tools like sonatype nexus since long time ago.

**Comment by jkroepke** [comment]
Hi @squillace,

is it possible to share to the roadmap which registry are going to implement next?

**Comment by squillace** [comment]
> Nice to see this feature. Right now, you need to configure each image you want to cache with a rule (up to 50) them the system do [this](https://learn.microsoft.com/en-us/azure/container-registry/buffer-gate-public-content?tabs=azure-cli#import-images-to-an-azure-container-registry) automatically for you.

Dear @leinad87, this initial release you can think of as "passive" or "manual" caching, in that you must configure a rule per repo -- not per tag/digest -- and then you manually pull in those images you want. The next time you pull an image from the cache, if that happens more than five minutes after the preceding pull, the Caching feature of ACR will go and check to see whether the tag has been updated and will pull in the newest version of that tag.

However, over the _next five months_ we will be adding some automation to this feature. There are some customer conflicts here; many customers just want to pay for an entire registry mirror and others do not want to pay for anything they do not use. Some want new images every five seconds; others are fine with once a day or even a week. The automatic "artifact sync" features, then, require some thinking to ensure that everyone gets as close to the specific behavior they want without having to wade through five screens of options and choices. 

Up to now, no one has requested "otherwise search the image on external repositories and cache it in the CR for future downloads" yet. Typically, the request is something like, "once I've asked to cache a set of tags, please check for me and update the cached images whether I've pulled them or not". 

And when you match that with things like retention policies, things start to get complicated fast. Now we have to resolve retention policies vs. automatic cached image hydration! 

All that said, you should be looking at something that will do what you want this next few months.

**Comment by squillace** [comment]
> Hi @squillace,
> 
> is it possible to share to the roadmap which registry are going to implement next?

YES. :-) We'll try to paste that in here or on the roadmap this week. We are just about locked and are already firing in this area.

**Comment by SimonHarrisonSH** [comment]
This is looking good! Another vote for Jfrog Artifactory, here's hoping it will appear high up the promised roadmap ;)

**Comment by jkroepke** [comment]
> YES. :-) We'll try to paste that in here or on the roadmap this week. We are just about locked and are already firing in this area.

Hi @squillace , where do you post it?

**Comment by hho** [comment]
@jkroepke The ACR public roadmap is [here](https://github.com/orgs/Azure/projects/259/views/1)

**Comment by MXClyde** [comment]
A customer of mine needs this feature for Sonatype Nexus repositories, where would be a good place to feature request this?

**Comment by thgossler** [comment]
> > Nice to see this feature. Right now, you need to configure each image you want to cache with a rule (up to 50) them the system do [this](https://learn.microsoft.com/en-us/azure/container-registry/buffer-gate-public-content?tabs=azure-cli#import-images-to-an-azure-container-registry) automatically for you.
> 
> Dear @leinad87, this initial release you can think of as "passive" or "manual" caching, in that you must configure a rule per repo -- not per tag/digest -- and then you manually pull in those images you want. The next time you pull an image from the cache, if that happens more than five minutes after the preceding pull, the Caching feature of ACR will go and check to see whether the tag has been updated and will pull in the newest version of that tag.
> 
> However, over the _next five months_ we will be adding some automation to this feature. There are some customer conflicts here; many customers just want to pay for an entire registry mirror and others do not want to pay for anything they do not use. Some want new images every five seconds; others are fine with once a day or even a week. The automatic "artifact sync" features, then, require some thinking to ensure that everyone gets as close to the specific behavior they want without having to wade through five screens of options and choices.
> 
> Up to now, no one has requested "otherwise search the image on external repositories and cache it in the CR for future downloads" yet. Typically, the request is something like, "once I've asked to cache a set of tags, please check for me and update the cached images whether I've pulled them or not".
> 
> And when you match that with things like retention policies, things start to get complicated fast. Now we have to resolve retention policies vs. automatic cached image hydration!
> 
> All that said, you should be looking at something that will do what you want this next few months.

I believe a generic pull-through feature should nicely play together with the retention policies. If a pulled-through image (tag/digest) from an external private registry (e.g. a self-hosted cloud instance of JFrog Container Registry) was not used for a certain period of time it would be removed. If it is accessed again later it would be pulled-through again.

We would like to setup a bidirectional pull-through between our Azure CR and JFrog CR (used for different purposes). In JFrog Artifactory we can setup the ACR as a Remote Repository (see: https://jfrog.com/help/r/jfrog-artifactory-documentation/jfrog-container-registry). On ACR side this seems not possible currently.

It would be great to have support for remote repositories in ACR as well.

**Comment by jmwoloso** [comment]
looking forward to this feature. how are we looking for late July with Nvidia?

https://github.com/orgs/Azure/projects/259/views/1?pane=issue&itemId=25500594

**Comment by davidkarlsen** [comment]
It is annoying that you need to define a cache-rule for each and every repo - having a wildcard like `docker.io/*` mapping over to a prefix of `docker.io/` within ACR would scale a lot better.

**Comment by davegallant** [comment]
> It is annoying that you need to define a cache-rule for each and every repo - having a wildcard like `docker.io/*` mapping over to a prefix of `docker.io/` within ACR would scale a lot better.

Yes, this adds a lot of tedium and makes this feature barely usable.

Artifactory has offered this functionality for a long time.

**Comment by SimonHarrisonSH** [comment]
> I believe a generic pull-through feature should nicely play together with the retention policies. If a pulled-through image (tag/digest) from an external private registry (e.g. a self-hosted cloud instance of JFrog Container Registry) was not used for a certain period of time it would be removed. If it is accessed again later it would be pulled-through again.
> 
> We would like to setup a bidirectional pull-through between our Azure CR and JFrog CR (used for different purposes). In JFrog Artifactory we can setup the ACR as a Remote Repository (see: https://jfrog.com/help/r/jfrog-artifactory-documentation/jfrog-container-registry). On ACR side this seems not possible currently.
> 
> It would be great to have support for private remote repositories with authentication in ACR as well.

This, exactly, that's all I'm looking for - a pull-through cache from Artifactory (just like Artifactory does). At the moment, we have to manually push the required images to ACR and manually clean-up.

**Comment by mllab-nl** [comment]
Nice to see * rules work now! 

Stange thing is why this works 
**docker pull myregistry.azurecr.io/docker.io/grafana/loki:2.8.2**
and this one 
**docker pull myregistry.azurecr.io/docker.io/alpine**
gives the error: _Error response from daemon: unauthorized: authentication required with source registry for cache rule 'docker-io' For more information, please visit https://aka.ms/acr/cache._

Do you guys have any idea why this is the case ?

**Comment by davidkarlsen** [comment]
Try... /library/alpine instead

**Comment by mllab-nl** [comment]
Thank you @davidkarlsen! 
**docker pull myregistry.azurecr.io/docker.io/library/alpine**
I now things are a bit more clear.

**Comment by JXavierMSFT** [comment]
> > It is annoying that you need to define a cache-rule for each and every repo - having a wildcard like `docker.io/*` mapping over to a prefix of `docker.io/` within ACR would scale a lot better.
> 
> Yes, this adds a lot of tedium and makes this feature barely usable.
> 
> Artifactory has offered this functionality for a long time.

We are currently planning on rolling out Wildcard capabilities. I will update this thread once they are released.

**Comment by davidkarlsen** [comment]
> > Yes, this adds a lot of tedium and makes this feature barely usable.
> > Artifactory has offered this functionality for a long time.
> 
> We are currently planning on rolling out Wildcard capabilities. I will update this thread once they are released.

Wow - it seems to work using * on source and target - but I cannot see any documentation on this?

**Comment by AliakseiVenski** [comment]
Wildcards work for me as well, @jxaviermsft, was it rolled out?

**Comment by JXavierMSFT** [comment]
Wildcards now are officially released! You can view the documentation on aka.ms/acr/cache. Please let me know if you have any feedback on the use of the feature.

**Comment by JXavierMSFT** [comment]
My apologies for the late reply @AliakseiVenski. I was out of the office for a while. We officially rolled out Wildcards to all public regions yesterday. You can check the docs at aka.ms/acr/cache.

**Comment by P4R4DiSi4C** [comment]
Thanks @JXavierMSFT and everyone that worked on this, I was able to quickly test it and it's wonderful !
I just encountered an issue and not sure where I can report it...In short, if you create credentials in ACR Cache that need access to a KeyVault, the RBAC won't find the ObjectID of that CredentialsSet (it works with AzCli tho...). The only way to grant RBAC access to the keyvault for ACR's CredentialsSet through Azure Portal GUI, is to use the generated Enterprise App fullname.

**Comment by ericyew** [comment]
what is happening with jfrog support as per [https://github.com/Azure/acr/issues/683](https://github.com/Azure/acr/issues/683)?

**Comment by emibcn** [comment]
I am trying the cache/pull through on a newly created ACR and I can see there are some inconsistencies.

## "Everything works as expected" only if...
... the cache resource is created with only the same target path as the source one. I'm using Terraform for creating it, but I believe you will be able to understand that:

### This works
If I set the Terraform cache resource to this values:
```hcl
resource "azurerm_container_registry_cache_rule" "docker-io" {                                                        
  name                  = "docker-io"
  container_registry_id = azurerm_container_registry.acr.id                                                           
  target_repo           = "*"
  source_repo           = "docker.io/*"
} 
```

And the Docker `daemon.json` to this:
```json
{
    "debug": true,
    "registry-mirrors": ["https://*************.azurecr.io"]
}
```

**Note:** I have hidden the ACR name: it's not `*************` ;)

### This does not work
If I set the Terraform cache resource to this values:
```hcl
resource "azurerm_container_registry_cache_rule" "docker-io" {                                                        
  name                  = "docker-io"
  container_registry_id = azurerm_container_registry.acr.id                                                           
  target_repo           = "docker-io/*" # Note I set a target path different from the source
  source_repo           = "docker.io/*"
} 
```

And the Docker `daemon.json` to this:
```json
{
    "debug": true,
    "registry-mirrors": ["https://*************.azurecr.io/docker-io"]
}
```

Docker daemon logs (using `journalctl -feu docker`):
```
level=debug msg="Trying to pull python from https://*************.azurecr.io/docker-io/" spanID=a6fa10a698a0acc7 traceID=f066d35980ab247b230e6b2a93e2364b
level=debug msg="Trying to pull python from https://*************.azurecr.io/docker-io/" spanID=a6fa10a698a0acc7 traceID=f066d35980ab247b230e6b2a93e2364b
level=info msg="Attempting next endpoint for pull after error: unknown: Not found" spanID=a6fa10a698a0acc7 traceID=f066d35980ab247b230e6b2a93e2364b
level=debug msg="Trying to pull python from https://registry-1.docker.io" spanID=a6fa10a698a0acc7 traceID=f066d35980ab247b230e6b2a93e2364b
level=debug msg="Trying to pull python from https://registry-1.docker.io" spanID=a6fa10a698a0acc7 traceID=f066d35980ab247b230e6b2a93e2364b
level=debug msg="Fetching manifest from remote" digest="sha256:4584ea46d313a10e849eb7c5ef36be14773418233516ceaa9e52a8ff7d5e35a5" error="<nil>" remote="docker.io/library/python:3.12" spanID=a6fa10a698a0acc7 traceID=f066d35980ab247b230e6b2a93e2364b
```

**Comment by davidkarlsen** [comment]
@emibcn same here - it used to work before though.

**Comment by Jamesits** [comment]
Wrote a small program to workaround the 2 problems:
- `registry-mirrors` does not work with a prefixed cache rule
- `registry-mirrors` does not support authentication at all https://github.com/moby/moby/issues/30880

https://github.com/Jamesits/acr-cache-proxy
